deepspeed/utils/zero_to_fp32.py

```diff
@@ -435,10 +496,9 @@ def _zero3_merge_trainable_params(state_dict, world_size, fp32_flat_groups, zero
                 f"Trainable params: {total_params} {name} full shape: {shape} partition0 numel={partitioned_numel} partitioned_padding_numel={partitioned_padding_numel}"
             )
 
-        # XXX: memory usage doubles here
-        state_dict[name] = torch.cat(
-            tuple(fp32_flat_groups[i].narrow(0, offset, partitioned_numel) for i in range(world_size)),
-            0).narrow(0, 0, unpartitioned_numel).view(shape)
+        # memory efficient tensor
+        tensor = GatheredTensor(fp32_flat_groups, offset, partitioned_numel, shape)
```

Comment:
This type change from `torch.Tensor` to a pseudo tensor seems problematic, especially for non-HF usage. Can you please explain the rationale and implications?

Reply:
Okay. now I see that pseudo tensor is converted to `torch.Tensor` by `contiguous()`.

Reply:
more comments have been added to the script.